File: frontend/src/lib/api/prefix-list.ts
import { apiClient } from "./client";

// ============================================================================
// TypeScript Interfaces
// ============================================================================

export interface PrefixListRule {
  rule_number: number;
  action: string;  // permit or deny
  description?: string | null;
  prefix?: string | null;  // CIDR notation (e.g., 10.0.0.0/8)
  ge?: number | null;  // Greater-than-or-equal prefix length
  le?: number | null;  // Less-than-or-equal prefix length
}

export interface PrefixList {
  name: string;
  description?: string | null;
  rules: PrefixListRule[];
  list_type: string;  // ipv4 or ipv6
}

export interface PrefixListConfigResponse {
  ipv4_lists: PrefixList[];
  ipv6_lists: PrefixList[];
  total_ipv4: number;
  total_ipv6: number;
}

export interface PrefixListCapabilitiesResponse {
  version: string;
  features: {
    ipv4_prefix_lists: {
      supported: boolean;
      description: string;
    };
    ipv6_prefix_lists: {
      supported: boolean;
      description: string;
    };
    ge_le_operators: {
      supported: boolean;
      description: string;
    };
    rule_descriptions: {
      supported: boolean;
      description: string;
    };
  };
  device_name?: string;
}

export interface PrefixListBatchOperation {
  op: string;
  value?: string;
}

export interface PrefixListBatchRequest {
  name: string;
  list_type: string;  // ipv4 or ipv6
  rule_number?: number;
  operations: PrefixListBatchOperation[];
}

// ============================================================================
// API Service
// ============================================================================

class PrefixListService {
  /**
   * Get capabilities based on VyOS version
   */
  async getCapabilities(): Promise<PrefixListCapabilitiesResponse> {
    return apiClient.get<PrefixListCapabilitiesResponse>("/vyos/prefix-list/capabilities");
  }

  /**
   * Get all prefix-list configurations
   */
  async getConfig(refresh: boolean = false): Promise<PrefixListConfigResponse> {
    return apiClient.get<PrefixListConfigResponse>("/vyos/prefix-list/config", {
      refresh: refresh.toString(),
    });
  }

  /**
   * Refresh the cached configuration
   */
  async refreshConfig(): Promise<unknown> {
    return apiClient.post("/vyos/config/refresh");
  }

  /**
   * Execute batch operations
   */
  async batchConfigure(request: PrefixListBatchRequest): Promise<unknown> {
    const result = await apiClient.post("/vyos/prefix-list/batch", request);
    await this.refreshConfig();
    return result;
  }

  /**
   * Reorder rules in a prefix-list
   */
  async reorderRules(
    name: string,
    listType: string,
    rules: Array<{ old_number: number; new_number: number; rule_data: PrefixListRule }>
  ): Promise<unknown> {
    const result = await apiClient.post("/vyos/prefix-list/reorder", {
      name,
      list_type: listType,
      rules,
    });
    await this.refreshConfig();
    return result;
  }

  /**
   * Helper: Create a new prefix-list with first rule
   */
  async createPrefixList(
    name: string,
    listType: string,
    description: string | null,
    rule: Partial<PrefixListRule>
  ): Promise<unknown> {
    const operations: PrefixListBatchOperation[] = [];

    // Create prefix-list
    operations.push({ op: listType === "ipv4" ? "set_prefix_list" : "set_prefix_list6" });

    // Add description if provided
    if (description) {
      operations.push({
        op: listType === "ipv4" ? "set_prefix_list_description" : "set_prefix_list6_description",
        value: description
      });
    }

    // Create rule
    operations.push({ op: listType === "ipv4" ? "set_rule" : "set_rule6" });

    // Set rule action
    operations.push({
      op: listType === "ipv4" ? "set_rule_action" : "set_rule6_action",
      value: rule.action || "permit"
    });

    // Set rule description
    if (rule.description) {
      operations.push({
        op: listType === "ipv4" ? "set_rule_description" : "set_rule6_description",
        value: rule.description
      });
    }

    // Set prefix
    if (rule.prefix) {
      operations.push({
        op: listType === "ipv4" ? "set_rule_prefix" : "set_rule6_prefix",
        value: rule.prefix
      });
    }

    // Set ge
    if (rule.ge !== undefined && rule.ge !== null) {
      operations.push({
        op: listType === "ipv4" ? "set_rule_ge" : "set_rule6_ge",
        value: rule.ge.toString()
      });
    }

    // Set le
    if (rule.le !== undefined && rule.le !== null) {
      operations.push({
        op: listType === "ipv4" ? "set_rule_le" : "set_rule6_le",
        value: rule.le.toString()
      });
    }

    return this.batchConfigure({
      name,
      list_type: listType,
      rule_number: rule.rule_number,
      operations,
    });
  }

  /**
   * Helper: Update prefix-list description
   */
  async updatePrefixListDescription(
    name: string,
    listType: string,
    description: string | null
  ): Promise<unknown> {
    const operations: PrefixListBatchOperation[] = [];

    if (description) {
      operations.push({
        op: listType === "ipv4" ? "set_prefix_list_description" : "set_prefix_list6_description",
        value: description
      });
    } else {
      operations.push({
        op: listType === "ipv4" ? "delete_prefix_list_description" : "delete_prefix_list6_description"
      });
    }

    return this.batchConfigure({
      name,
      list_type: listType,
      operations,
    });
  }

  /**
   * Helper: Delete a prefix-list
   */
  async deletePrefixList(name: string, listType: string): Promise<unknown> {
    const operations: PrefixListBatchOperation[] = [];
    operations.push({
      op: listType === "ipv4" ? "delete_prefix_list" : "delete_prefix_list6"
    });

    return this.batchConfigure({
      name,
      list_type: listType,
      operations,
    });
  }

  /**
   * Helper: Add a rule to an existing prefix-list
   */
  async addRule(
    name: string,
    listType: string,
    rule: Partial<PrefixListRule>
  ): Promise<unknown> {
    const operations: PrefixListBatchOperation[] = [];

    // Create rule
    operations.push({ op: listType === "ipv4" ? "set_rule" : "set_rule6" });

    // Set rule action
    operations.push({
      op: listType === "ipv4" ? "set_rule_action" : "set_rule6_action",
      value: rule.action || "permit"
    });

    // Set rule description
    if (rule.description) {
      operations.push({
        op: listType === "ipv4" ? "set_rule_description" : "set_rule6_description",
        value: rule.description
      });
    }

    // Set prefix
    if (rule.prefix) {
      operations.push({
        op: listType === "ipv4" ? "set_rule_prefix" : "set_rule6_prefix",
        value: rule.prefix
      });
    }

    // Set ge
    if (rule.ge !== undefined && rule.ge !== null) {
      operations.push({
        op: listType === "ipv4" ? "set_rule_ge" : "set_rule6_ge",
        value: rule.ge.toString()
      });
    }

    // Set le
    if (rule.le !== undefined && rule.le !== null) {
      operations.push({
        op: listType === "ipv4" ? "set_rule_le" : "set_rule6_le",
        value: rule.le.toString()
      });
    }

    return this.batchConfigure({
      name,
      list_type: listType,
      rule_number: rule.rule_number,
      operations,
    });
  }

  /**
   * Helper: Update an existing rule
   */
  async updateRule(
    name: string,
    listType: string,
    ruleNumber: number,
    rule: Partial<PrefixListRule>
  ): Promise<unknown> {
    const operations: PrefixListBatchOperation[] = [];

    // IMPORTANT: Process deletes FIRST, then sets
    // This ensures VyOS properly removes fields before setting new values

    // Delete operations first
    if (rule.description !== undefined && !rule.description) {
      operations.push({
        op: listType === "ipv4" ? "delete_rule_description" : "delete_rule6_description"
      });
    }

    if (rule.ge !== undefined && rule.ge === null) {
      operations.push({
        op: listType === "ipv4" ? "delete_rule_ge" : "delete_rule6_ge"
      });
    }

    if (rule.le !== undefined && rule.le === null) {
      operations.push({
        op: listType === "ipv4" ? "delete_rule_le" : "delete_rule6_le"
      });
    }

    // Set operations after deletes
    if (rule.action) {
      operations.push({
        op: listType === "ipv4" ? "set_rule_action" : "set_rule6_action",
        value: rule.action
      });
    }

    if (rule.description !== undefined && rule.description) {
      operations.push({
        op: listType === "ipv4" ? "set_rule_description" : "set_rule6_description",
        value: rule.description
      });
    }

    if (rule.prefix !== undefined && rule.prefix) {
      operations.push({
        op: listType === "ipv4" ? "set_rule_prefix" : "set_rule6_prefix",
        value: rule.prefix
      });
    }

    if (rule.ge !== undefined && rule.ge !== null) {
      operations.push({
        op: listType === "ipv4" ? "set_rule_ge" : "set_rule6_ge",
        value: rule.ge.toString()
      });
    }

    if (rule.le !== undefined && rule.le !== null) {
      operations.push({
        op: listType === "ipv4" ? "set_rule_le" : "set_rule6_le",
        value: rule.le.toString()
      });
    }

    return this.batchConfigure({
      name,
      list_type: listType,
      rule_number: ruleNumber,
      operations,
    });
  }

  /**
   * Helper: Delete a rule
   */
  async deleteRule(
    name: string,
    listType: string,
    ruleNumber: number
  ): Promise<unknown> {
    const operations: PrefixListBatchOperation[] = [];
    operations.push({
      op: listType === "ipv4" ? "delete_rule" : "delete_rule6"
    });

    return this.batchConfigure({
      name,
      list_type: listType,
      rule_number: ruleNumber,
      operations,
    });
  }
}

export const prefixListService = new PrefixListService();

File: backend/routers/prefix_list/prefix_list.py
"""
Prefix List Router

API endpoints for managing VyOS prefix-list configuration.
Supports both IPv4 (prefix-list) and IPv6 (prefix-list6) with version-aware configuration.
"""

from fastapi import APIRouter, HTTPException, Request
from starlette.concurrency import run_in_threadpool
from pydantic import BaseModel, Field
from typing import List, Dict, Optional, Any
from session_vyos_service import get_session_vyos_service
from vyos_builders import PrefixListBatchBuilder
from fastapi_permissions import require_read_permission, require_write_permission
from rbac_permissions import FeatureGroup
from utils.router_helpers import load_vyos_capabilities
import inspect

router = APIRouter(prefix="/vyos/prefix-list", tags=["prefix-list"])

# Stub functions for backwards compatibility with app.py
def set_device_registry(registry):
    """Legacy function - no longer used."""
    pass


def set_configured_device_name(name):
    """Legacy function - no longer used."""
    pass


# ============================================================================
# Pydantic Models
# ============================================================================


class PrefixListRule(BaseModel):
    """Model for a prefix-list rule."""
    rule_number: int
    action: str  # permit or deny
    description: Optional[str] = None
    prefix: Optional[str] = None  # CIDR notation (e.g., 10.0.0.0/8)
    ge: Optional[int] = None  # Greater-than-or-equal prefix length
    le: Optional[int] = None  # Less-than-or-equal prefix length


class PrefixList(BaseModel):
    """Model for a prefix-list."""
    name: str
    description: Optional[str] = None
    rules: List[PrefixListRule] = []
    list_type: str  # ipv4 or ipv6


class PrefixListConfigResponse(BaseModel):
    """Response containing all prefix-list configuration."""
    ipv4_lists: List[PrefixList] = []
    ipv6_lists: List[PrefixList] = []
    total_ipv4: int = 0
    total_ipv6: int = 0


class PrefixListBatchOperation(BaseModel):
    """Single operation in a batch request."""
    op: str = Field(..., description="Operation name")
    value: Optional[str] = Field(None, description="Operation value")


class PrefixListBatchRequest(BaseModel):
    """Model for batch configuration."""
    name: str = Field(..., description="Prefix list name")
    list_type: str = Field(..., description="ipv4 or ipv6")
    rule_number: Optional[int] = Field(None, description="Rule number for rule operations")
    operations: List[PrefixListBatchOperation]


class ReorderRuleItem(BaseModel):
    """Model for a single rule in reorder request."""
    old_number: int
    new_number: int
    rule_data: PrefixListRule


class ReorderPrefixListRequest(BaseModel):
    """Model for reordering rules in a prefix-list."""
    name: str = Field(..., description="Prefix list name")
    list_type: str = Field(..., description="ipv4 or ipv6")
    rules: List[ReorderRuleItem]


class VyOSResponse(BaseModel):
    """Standard response from VyOS operations."""
    success: bool
    data: Optional[Dict[str, Any]] = None
    error: Optional[str] = None


# ============================================================================
# Endpoint 1: Capabilities
# ============================================================================


@router.get("/capabilities")
async def get_prefix_list_capabilities(request: Request):
    """
    Get feature capabilities based on device VyOS version.

    Returns feature flags indicating which operations are supported.
    Allows frontends to conditionally enable/disable features.
    """
    return await load_vyos_capabilities(
        request=request,
        feature_group=FeatureGroup.PREFIX_LIST,
        builder_cls=PrefixListBatchBuilder,
    )


# ============================================================================
# Endpoint 2: Config (Generalized Data)
# ============================================================================


@router.get("/config", response_model=PrefixListConfigResponse)
async def get_prefix_list_config(http_request: Request, refresh: bool = False):
    """
    Get all prefix-list configurations from VyOS in a generalized format.

    Args:
        refresh: If True, force refresh from VyOS. If False, use cache.

    Returns:
        Generalized configuration data optimized for frontend consumption
    """
    # Check RBAC permission
    await require_read_permission(http_request, FeatureGroup.PREFIX_LIST)

    try:
        service = get_session_vyos_service(http_request)
        full_config = await run_in_threadpool(service.get_full_config, refresh=refresh)

        ipv4_lists = []
        ipv6_lists = []

        # Parse IPv4 prefix-lists
        policy_config = full_config.get("policy", {})
        prefix_lists = policy_config.get("prefix-list", {})

        for name, list_config in prefix_lists.items():
            rules = []
            rules_config = list_config.get("rule", {})

            for rule_num, rule_config in rules_config.items():
                # Parse rule data
                action = rule_config.get("action", "permit")
                description = rule_config.get("description")
                prefix = rule_config.get("prefix")

                # Parse ge (greater-than-or-equal)
                ge = None
                ge_value = rule_config.get("ge")
                if ge_value is not None:
                    try:
                        ge = int(ge_value)
                    except (ValueError, TypeError):
                        pass

                # Parse le (less-than-or-equal)
                le = None
                le_value = rule_config.get("le")
                if le_value is not None:
                    try:
                        le = int(le_value)
                    except (ValueError, TypeError):
                        pass

                rule = PrefixListRule(
                    rule_number=int(rule_num),
                    action=action,
                    description=description,
                    prefix=prefix,
                    ge=ge,
                    le=le,
                )
                rules.append(rule)

            # Sort rules by rule number
            rules.sort(key=lambda r: r.rule_number)

            prefix_list = PrefixList(
                name=name,
                description=list_config.get("description"),
                rules=rules,
                list_type="ipv4",
            )
            ipv4_lists.append(prefix_list)

        # Parse IPv6 prefix-lists
        prefix_lists6 = policy_config.get("prefix-list6", {})

        for name, list_config in prefix_lists6.items():
            rules = []
            rules_config = list_config.get("rule", {})

            for rule_num, rule_config in rules_config.items():
                # Parse rule data
                action = rule_config.get("action", "permit")
                description = rule_config.get("description")
                prefix = rule_config.get("prefix")

                # Parse ge (greater-than-or-equal)
                ge = None
                ge_value = rule_config.get("ge")
                if ge_value is not None:
                    try:
                        ge = int(ge_value)
                    except (ValueError, TypeError):
                        pass

                # Parse le (less-than-or-equal)
                le = None
                le_value = rule_config.get("le")
                if le_value is not None:
                    try:
                        le = int(le_value)
                    except (ValueError, TypeError):
                        pass

                rule = PrefixListRule(
                    rule_number=int(rule_num),
                    action=action,
                    description=description,
                    prefix=prefix,
                    ge=ge,
                    le=le,
                )
                rules.append(rule)

            # Sort rules by rule number
            rules.sort(key=lambda r: r.rule_number)

            prefix_list = PrefixList(
                name=name,
                description=list_config.get("description"),
                rules=rules,
                list_type="ipv6",
            )
            ipv6_lists.append(prefix_list)

        # Sort prefix lists by name
        ipv4_lists.sort(key=lambda pl: pl.name)
        ipv6_lists.sort(key=lambda pl: pl.name)

        return PrefixListConfigResponse(
            ipv4_lists=ipv4_lists,
            ipv6_lists=ipv6_lists,
            total_ipv4=len(ipv4_lists),
            total_ipv6=len(ipv6_lists),
        )
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


# ============================================================================
# Endpoint 3: Batch Operations
# ============================================================================


@router.post("/batch")
async def prefix_list_batch_configure(http_request: Request, body: PrefixListBatchRequest):
    """
    Execute a batch of configuration operations.

    Allows multiple changes in a single VyOS commit for efficiency.
    """
    # Check RBAC permission
    await require_write_permission(http_request, FeatureGroup.PREFIX_LIST)

    try:
        service = get_session_vyos_service(http_request)
        version = service.get_version()
        builder = PrefixListBatchBuilder(version=version)

        # Process operations using inspect for dynamic method calls
        for operation in body.operations:
            method = getattr(builder, operation.op)
            sig = inspect.signature(method)
            params = list(sig.parameters.keys())

            # Build arguments dynamically
            args = []

            # Add name
            if "name" in params:
                args.append(body.name)

            # Add rule number if present in method signature
            if "rule" in params and body.rule_number is not None:
                args.append(str(body.rule_number))

            # Add operation value
            if operation.value and len(params) > len(args):
                args.append(operation.value)

            method(*args)

        # Execute batch
        response = service.execute_batch(builder)

        return VyOSResponse(
            success=response.status == 200,
            data={"message": "Configuration updated"},
            error=response.error if response.error else None
        )
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


# ============================================================================
# Reorder Endpoint
# ============================================================================


@router.post("/reorder")
async def reorder_prefix_list_rules(http_request: Request, body: ReorderPrefixListRequest):
    """
    Reorder rules in a prefix-list.

    This endpoint deletes all existing rules and recreates them with new numbers
    in a single VyOS commit for atomicity.
    """
    # Check RBAC permission
    await require_write_permission(http_request, FeatureGroup.PREFIX_LIST)

    try:
        service = get_session_vyos_service(http_request)
        version = service.get_version()
        builder = PrefixListBatchBuilder(version=version)

        # Step 1: Delete all rules in reverse order
        rules_to_delete = sorted(
            [item.old_number for item in body.rules],
            reverse=True
        )

        for old_number in rules_to_delete:
            if body.list_type == "ipv4":
                builder.delete_rule(body.name, str(old_number))
            else:  # ipv6
                builder.delete_rule6(body.name, str(old_number))

        # Step 2: Recreate rules with new numbers
        for item in body.rules:
            rule = item.rule_data
            rule_str = str(item.new_number)

            if body.list_type == "ipv4":
                # Create rule
                builder.set_rule(body.name, rule_str)

                # Set action
                builder.set_rule_action(body.name, rule_str, rule.action)

                # Set description
                if rule.description:
                    builder.set_rule_description(
                        body.name, rule_str, rule.description
                    )

                # Set prefix
                if rule.prefix:
                    builder.set_rule_prefix(body.name, rule_str, rule.prefix)

                # Set ge
                if rule.ge is not None:
                    builder.set_rule_ge(body.name, rule_str, str(rule.ge))

                # Set le
                if rule.le is not None:
                    builder.set_rule_le(body.name, rule_str, str(rule.le))

            else:  # ipv6
                # Create rule
                builder.set_rule6(body.name, rule_str)

                # Set action
                builder.set_rule6_action(body.name, rule_str, rule.action)

                # Set description
                if rule.description:
                    builder.set_rule6_description(
                        body.name, rule_str, rule.description
                    )

                # Set prefix
                if rule.prefix:
                    builder.set_rule6_prefix(body.name, rule_str, rule.prefix)

                # Set ge
                if rule.ge is not None:
                    builder.set_rule6_ge(body.name, rule_str, str(rule.ge))

                # Set le
                if rule.le is not None:
                    builder.set_rule6_le(body.name, rule_str, str(rule.le))

        # Execute batch
        response = service.execute_batch(builder)

        return VyOSResponse(
            success=response.status == 200,
            data={"message": f"Rules reordered in prefix-list {body.name}"},
            error=response.error if response.error else None
        )
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))
